Sources/xcodeproj/Utils/PBXBatchUpdater.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+import PathKit
+
+// This is a helper class for quickly adding a large number of files.
+// It is forbidden to add a file to a group one by one using the PBXGroup method addFile(...) while you are working with this class.
+public final class PBXBatchUpdater {
+    private let projectObjects: PBXObjects
+    private let sourceRoot: Path
+    private var references: [Path: PBXObjectReference]?
+
+    init(projectObjects: PBXObjects, sourceRoot: Path) {
+        self.projectObjects = projectObjects
+        self.sourceRoot = sourceRoot
+    }
+
+    /// Adds file at the give path to the project or returns existing file and its reference.
+    ///
+    /// - Parameters:
+    ///   - fileName: file name.
+    ///   - sourceTree: file sourceTree, default is `.group`
+    ///   - sourceRoot: path to project's source root.
+    /// - Returns: new or existing file and its reference.
+    @discardableResult
+    public func addFile(
+        to group: PBXGroup,
+        fileName: String,
+        sourceTree: PBXSourceTree = .group
+    ) throws -> PBXFileReference {
+        let groupPath = try group.fullPath(sourceRoot: sourceRoot)
+        let filePath = (groupPath ?? sourceRoot) + Path(fileName)
+
+        guard filePath.exists else {
```

Comment:
Aligned with @mxcl's suggestion, we can make the API throw less by assuming the developer knows that it's doing. I'd expect the user of the API to check whether the file exists or not. If we build on that assumption, I don't think we need the check here.

Reply:
I fully agree, this test should not be here. For Xcode, this is a valid case to have a reference to the file, but the file itself may not exist.

I also need to remove this check in the PBXGroup addFile method.